src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+        # to zero, by subtracting the fixed value.
+
+        _value = np.atleast_1d(np.array(c["value"], dtype=float))
+
+        def constraint_from_internal(x):
+            params = converter.params_from_internal(x)
+            select = external_selector(params)
+            out = np.atleast_1d(constraint_func(select)) - _value
+            return out
+
+        jacobian_from_internal = _jacobian_from_internal
+        n_constr = _n_constr
+
+    else:
+
+        # ==============================================================================
+        # Inequality constraints
+        #
+        # We define the internal constraint function to be satisfied if it is
+        # greater than or equal to zero (positivity constraint). If the bounds already
+        # satify this condition we do not change anything, otherwise we need to perform
+        # a transformation.
+
+        def _constraint_from_internal(x):
+            params = converter.params_from_internal(x)
+            select = external_selector(params)
+            return np.atleast_1d(constraint_func(select))
+
+        lower_bounds = c.get("lower_bounds", 0)
+        upper_bounds = c.get("upper_bounds", np.inf)
+
+        transformation = _get_transformation(lower_bounds, upper_bounds)
+
+        constraint_from_internal = _compose_funcs(
+            _constraint_from_internal, transformation["func"]
+        )
+
+        jacobian_from_internal = _compose_funcs(
+            _jacobian_from_internal, transformation["derivative"]
+        )
+
+        n_constr = 2 * _n_constr if transformation["name"] == "stack" else _n_constr
+
+    internal_constr = {
+        "n_constr": n_constr,
+        "type": _type,
+        "fun": constraint_from_internal,  # internal name for 'func'
+        "jac": jacobian_from_internal,  # internal name for 'derivative'
+        "tol": c.get("tol", CONVERGENCE_ABSOLUTE_CONSTRAINT_TOLERANCE),
+    }
+
+    return internal_constr
+
+
+# ======================================================================================
+# Transform equality constraints to inequality constraints
+# ======================================================================================
+
+
+def equality_as_inequality_constraints(nonlinear_constraints):
+    """Return constraints where equality constraints are converted to inequality."""
+    constraints = [_equality_to_inequality(c) for c in nonlinear_constraints]
+    return constraints
+
+
+def _equality_to_inequality(c):
+    """Transform a single constraint.
+
+    An equality constaint g(x) = 0 can be transformed to two inequality constraints
+    using (g(x), -g(x)) >= 0. Hence, the number of constraints doubles, and the
+    constraint functions itself as well as the derivative need to be updated.
+
+    """
+    if c["type"] == "eq":
+
+        def transform(x, func):
+            return np.concatenate((func(x), -func(x)), axis=0)
```

Comment:
You can evaluate `func` once and then stack it with its negative. `func` can be slow!

Reply:
Very very good catch!